### ISSUE TITLE
support retail ThreeSlice

### ISSUE THREAD

**Issue by ferronn-dev** [ferronn-dev/wowless#34]
Our story begins with this stacktrace:

```
error: [string "SharedUIPanelTemplates.lua"]:1217: attempt to concatenate field 'atlasName' (a nil value)
stack traceback:
        ./wowless/api.lua:47: in function <./wowless/api.lua:45>
        [string "SharedUIPanelTemplates.lua"]:1217: in function 'GetLeftAtlasName'
        [string "SharedUIPanelTemplates.lua"]:1233: in function 'InitButton'
        [string "SharedUIPanelTemplates.lua"]:1330: in function '?'
        ./wowless/loader.lua:184: in function 'script'
        ./wowless/api.lua:59: in function <./wowless/api.lua:59>
```

The GetLeftAtlasName call looks like this:

```lua
ThreeSliceButtonMixin = CreateFromMixins(UIButtonMixin);
function ThreeSliceButtonMixin:GetLeftAtlasName()
	return self.atlasName.."-Left";
end
```

The InitButton call is happening via OnLoad of the controller.

```lua
-- Allows inheriting buttons to override OnLoad and OnShow
ButtonControllerMixin = {};
function ButtonControllerMixin:OnLoad()
	if self:GetParent().InitButton then
		self:GetParent():InitButton();
	end
end
```

The error is being triggered by the instantiation of AddonListCancelButton:

```xml
<Button name="$parentCancelButton" inherits="SharedButtonSmallTemplate" parentKey="CancelButton" text="CANCEL">
```

which is a

```xml
<Button name="SharedButtonSmallTemplate" inherits="BigRedThreeSliceButtonTemplate" virtual="true">
```

which is a

```xml
  <Button name="BigRedThreeSliceButtonTemplate" inherits="ThreeSliceButtonTemplate" virtual="true">
    <Size x="441" y="128"/>
    <KeyValues>
      <KeyValue key="atlasName" value="128-RedButton" type="string"/>
    </KeyValues>
    <PushedTextOffset x="-2" y="-1"/>
  </Button>
```

which is a

```xml
  <Button name="ThreeSliceButtonTemplate" mixin="ThreeSliceButtonMixin" motionScriptsWhileDisabled="true" virtual="true">
    <!--
    atlasName: The art kit to use for the 3 sections of the button. 
            * The left textures will be "atlasName-Left", "atlasName-Left-Disabled" and "atlasName-Left-Pressed"
            * The right textures will be "atlasName-Right", "atlasName-Right-Disabled" and "atlasName-Right-Pressed"
            * The center textures will be "_atlasName-Center", "_atlasName-Disabled" and "_atlasName-Pressed"
            * The highlight texture will be "atlasName-Highlight"
    tooltip: This can be either a string or a function which returns a string
 
    <KeyValues>
      <KeyValue key="atlasName" value="128-RedButton" type="string"/>
      <KeyValue key="tooltip" value="MyTooltipFunction" type="global"/>
        or
      <KeyValue key="tooltip" value="REFRESH_BUTTON_TOOLTIP" type="string"/>
    </KeyValues>
    -->
    <Size x="20" y="20"/>
    <Layers>
      <Layer level="BACKGROUND">
        <Texture parentKey="Left">
          <Anchors>
            <Anchor point="TOPLEFT"/>
          </Anchors>
        </Texture>
        <Texture parentKey="Right">
          <Anchors>
            <Anchor point="TOPRIGHT"/>
          </Anchors>
        </Texture>
        <Texture parentKey="Center" horizTile="true">
          <Anchors>
            <Anchor point="TOPLEFT" relativeKey="$parent.Left" relativePoint="TOPRIGHT"/>
            <Anchor point="BOTTOMRIGHT" relativeKey="$parent.Right" relativePoint="BOTTOMLEFT"/>
          </Anchors>
        </Texture>
      </Layer>
    </Layers>
    <Frames>
      <Frame parentKey="Controller" mixin="ButtonControllerMixin">
        <Scripts>
          <OnLoad method="OnLoad"/>
          <OnShow method="OnShow"/>
        </Scripts>
      </Frame>
    </Frames>
    <Scripts>
      <OnMouseDown method="OnMouseDown"/>
      <OnMouseUp method="OnMouseUp"/>
      <OnShow method="UpdateButton"/>
      <OnDisable method="UpdateButton"/>
      <OnEnable method="UpdateButton"/>
      <OnEnter method="OnEnter"/>
      <OnLeave method="OnLeave"/>
      <OnSizeChanged method="UpdateScale"/>
    </Scripts>
  </Button>
```

which has the dependency on atlasName, the controller, and the mixin where this all started.

The problem appears to be that the controller's OnLoad is running after its parent, AddonListCancelButton, has had its mixins applied (which brings in InitButton) but before its parent has had all its keyvalues applied (which brings in atlasName). The controller's OnLoad should probably run after the keyvalues are applied. This means that the keyvalues, which live on BigRedThreeSliceButtonTemplate, need to be applied before the children of ThreeSliceButtonTemplate (from which it inherits) are built.

In summary it seems that certain frame children (like keyvalues) must be applied through the entire inheritance chain before others (like frames).

**Comment by ferronn-dev** [comment]
Meo confirms that "static stuff" needs to be applied before dealing with child frames.

**Comment by ferronn-dev** [comment]
fixed in 52b40a1132f749a7be068767c806d139e858190b
